Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/FacepunchMultiplayerMenu/FacepunchMultiplayerMenu.cs

```diff
@@ -0,0 +1,210 @@
+
+	private NetworkManager mgr = null;
+	private NetWorker server;
+	private Steamworks.Data.Lobby lobby;
+	private List<Button> _uiButtons = new List<Button>();
+
+	private void Start()
+	{
+		for (int i = 0; i < toggledButtons.Length; ++i)
+		{
+			Button btn = toggledButtons[i].GetComponent<Button>();
+			if (btn != null)
+				_uiButtons.Add(btn);
+		}
+
+		if (useMainThreadManagerForRPCs)
+			Rpc.MainThreadRunner = MainThreadManager.Instance;
+	}
+
+	/// <summary>
+	/// Connects to the currently selected Steam Lobby
+	/// Lobby selection is made within FacepuncJoinMenu.cs
+	/// </summary>
+	public void Connect()
+	{
+		if (lobbyToJoin.Id > 0)
+		{
+			ConnectToLobbyAsync();
+		}
+
+		// TODO:  Raise a dialog to inform of the failed connection attempt
+	}
+
+	/// <summary>
+	/// Connect to selected steam lobby asynchronously
+	/// Void method calls the async task
+	/// </summary>
+	private async void ConnectToLobbyAsync()
+	{
+		await ConnectToLobby();
+	}
+
+	/// <summary>
+	/// Connect to the selected steam lobby
+	/// On successful connection, starts a new FacepunchP2PClient for connection to the lobby owner's FacepunchP2PServer
+	/// </summary>
+	private async Task ConnectToLobby()
+	{
+		RoomEnter roomEnter = await lobbyToJoin.Join();
+		if (roomEnter != RoomEnter.Success)
+		{
+			BMSLog.Log("Error connecting to lobby returned: " + roomEnter.ToString());
+			return;
+		}
+
+		this.lobby = lobbyToJoin;
+		ConnectToServer(lobbyToJoin.Owner.Id);
+	}
+
+	/// <summary>
+	/// Creates a new FacepunchP2PClient and attempts to connect to the specified SteamId
+	/// </summary>
+	/// <param name="hostSteamId">SteamId of the host/server</param>
+	private void ConnectToServer(SteamId hostSteamId)
+	{
+		var client = new FacepunchP2PClient();
+		client.Connect(hostSteamId);
+		Connected(client);
+	}
+
+	/// <summary>
+	/// Create a new Steam lobby and FacepunchP2PServer
+	/// </summary>
+	public void Host()
+	{
+		CreateLobbyAsync();
+	}
+
+	/// <summary>
+	/// Creates steam lobby asynchronously
+	/// Void method calls the async task
+	/// </summary>
+	private async void CreateLobbyAsync()
+	{
+		await CreateLobby();
+	}
+
+	/// <summary>
+	/// Create steam lobby asynchronously
+	/// Starts a new FacepunchP2PServer if lobby creation successful
+	/// </summary>
+	/// <returns></returns>
+	private async Task CreateLobby()
+	{
+		Steamworks.Data.Lobby? lobby = await SteamMatchmaking.CreateLobbyAsync(MAXIMUM_SERVER_SLOTS);
+		if (lobby.HasValue)
+		{
+			CreateFacepunchP2PServer();
+		}
+
+		// TODO:  Raise a dialog to inform of the failed lobby creation attempt
+	}
+
+	/// <summary>
+	/// Create new FacepunchP2PServer
+	/// </summary>
+	private void CreateFacepunchP2PServer()
+	{
+		server = new FacepunchP2PServer(MAXIMUM_SERVER_SLOTS);
+		((FacepunchP2PServer)server).Host();
+
+		server.playerTimeout += (player, sender) =>
+		{
+			BMSLog.Log("Player " + player.NetworkId + " timed out");
+		};
+
+		// TODO:  Add in server callbacks here as required
```

Comment:
Maybe add some generic events that users can subscribe to instead of requiring them to edit core Forge files.